File: TabsApp/app/src/main/java/ca/psdev/android/sandbox/tabsapp/MainActivity.java
package ca.psdev.android.sandbox.tabsapp;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.google.android.material.tabs.TabLayout;

import org.w3c.dom.Text;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        SectionsPagerAdapter adapter = new SectionsPagerAdapter(getSupportFragmentManager());
        ViewPager pager = (ViewPager) findViewById(R.id.pager);
        pager.setAdapter(adapter);

        TabLayout tabs = (TabLayout) findViewById(R.id.tabs);
        tabs.setupWithViewPager(pager);
    }

    public void onClickText(View view) {
        TextView textView = (TextView) findViewById(R.id.topDiplay);
        textView.setText("CLICKED MAIN ACTIVITY");
    }

    private class SectionsPagerAdapter extends FragmentPagerAdapter {

        public SectionsPagerAdapter(@NonNull FragmentManager fm) {
            super(fm, FragmentPagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        }

        @NonNull
        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case 0:
                    return new TopFragment();
                case 1:
                    return new ItemFragment();
                case 2:
                    return new DataFragment();
            }
            return null;
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case 0:
                    return "Home";
                case 1:
                    return "Items";
                case 2:
                    return "Data";
            }
            return null;
        }

        @Override
        public int getCount() {
            return 3;
        }
    }

}

File: ListView/app/src/main/java/ca/psdev/android/sandbox/listview/MainActivity.kt
package ca.psdev.android.sandbox.listview

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.BaseAdapter
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private var mItems = ArrayList<String>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar as Toolbar)


        for (i in 1..101) {
            mItems.add("Item # %03d".format(i))
        }

        lv_items.setOnItemClickListener { _, _, position, _ ->
            val item = mItems.get(position)
            Toast.makeText(this, "Clicked '$item'", Toast.LENGTH_SHORT).show()
        }

        // val adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, mItems)
        // lv_items.adapter = adapter
        lv_items.adapter = StringAdapter(this, mItems)
    }

    // get more info here - https://www.raywenderlich.com/155-android-listview-tutorial-with-kotlin
    class StringAdapter(private val context: Context,
                        private val data: ArrayList<String>) : BaseAdapter() {

        private val inflater: LayoutInflater = context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

        override fun getCount(): Int {
            return data.size
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
            val view = inflater.inflate(android.R.layout.simple_list_item_1, parent, false) as TextView
            view.text = data[position].toUpperCase()
            return view;
        }

        override fun getItem(position: Int): Any {
            return data.get(position).toUpperCase()
        }

        override fun getItemId(position: Int): Long {
            return position.toLong()
        }
    }

}
